### ISSUE TITLE
Add Type Discriminator to ZodType

### ISSUE THREAD

**Issue by rubber-duck-software** [colinhacks/zod#522]
Occasionally, users want to map a zod type to some other value. This problem comes up in mocking, getting defaults, generating ui from zod types, etc. Whenever a user is mapping zod types, it is useful to have an accurate typescript type for the result. For this to be possible, it seems to me that the typescript types for each option in ZodFirstPartySchemaTypes should be non-overlapping. This avoids situations like the following...

```js
import { z } from 'zod';

type myType = z.ZodAny extends z.ZodNull ? true : false; // true

const myValue = z.any() instanceof z.ZodNull; // false
```

Here, I believe ZodAny is meant to be a distinct type from ZodNull, but it is inferred as extending ZodNull.

My proposed fix is adding a type discriminator to the _def of all ZodFirstPartySchemaTypes. Something like, ```{ typeName: "ZodString" }```

**Comment by scolestock** [comment]
Agree.  Writing on behalf of https://github.com/Marcato-Partners.

**Comment by colinhacks** [comment]
This was an intentional decision in Zod 3 that enables greater extensibility of Zod. Now anyone can subclass `ZodType` to create new functionality not covered by first-party types.

I think the solution is to provide a function that accepts a schema, handles all the `instanceof` logic, and returns a discriminated union of first-party types: `{kind: "string": schema: ZodString} | {kind: "number"; schema: ZodNumber} | ...`. Sound  good?

**Comment by colinhacks** [comment]
Actually I suppose a discriminator could also be added back to all the first-party subclasses without sacrificing extensibility... 🤔  The real problem in Zod 1/2 was that all schema types shared a single global parsing function instead of encapsulating the parsing logic inside the subclass. Okay I'm down for this; feel free to submit a PR.

**Comment by colinhacks** [comment]
Merged in #523
